Storage/Rust/RustLogins.swift

```diff
@@ -373,35 +373,119 @@ public class RustLoginEncryptionKeys {
 
     func decryptSecureFields(login: EncryptedLogin) -> Login? {
-        guard let key = self.keychain.string(forKey: self.loginPerFieldKeychainKey) else { return nil }
+        guard let key = self.keychain.string(forKey: self.loginPerFieldKeychainKey) else {
+            return nil
+        }
 
         do {
             return try decryptLogin(login: login, encryptionKey: key)
         } catch let err as NSError {
-            SentryIntegration.shared.sendWithStacktrace(message: "Error decrypting login", tag: SentryTag.rustLogins, severity: .error, description: err.localizedDescription)
-            return nil
+            if let loginsStoreError = err as? LoginsStoreError {
+                sendLoginsStoreErrorToSentry(
+                    err: loginsStoreError,
+                    errDomain: err.domain,
+                    errorMessage: "Error while decrypting login")
+            } else {
+                SentryIntegration.shared.sendWithStacktrace(
+                    message: "Unknown error while decrypting login",
+                    tag: SentryTag.rustLogins,
+                    severity: .error,
+                    description: err.localizedDescription)
+            }
         }
+        return nil
     }
 
-    func encryptSecureFields(login: Login, encryptionKey: String? = nil) -> EncryptedLogin? {
-        guard let key = self.keychain.string(forKey: self.loginPerFieldKeychainKey) else { return nil }
+    func encryptSecureFields(
+        login: Login,
+        encryptionKey: String? = nil
+    ) -> EncryptedLogin? {
+        guard let key = self.keychain.string(forKey: self.loginPerFieldKeychainKey) else {
+            return nil
+        }
 
         do {
             return try encryptLogin(login: login, encryptionKey: key)
         } catch let err as NSError {
-            SentryIntegration.shared.sendWithStacktrace(message: "Error encrypting login", tag: SentryTag.rustLogins, severity: .error, description: err.localizedDescription)
-            return nil
+            if let loginsStoreError = err as? LoginsStoreError {
+                sendLoginsStoreErrorToSentry(
+                    err: loginsStoreError,
+                    errDomain: err.domain,
+                    errorMessage: "Error while encrypting login")
+            } else {
+                SentryIntegration.shared.sendWithStacktrace(
+                    message: "Unknown error while encrypting login",
+                    tag: SentryTag.rustLogins,
+                    severity: .error,
+                    description: err.localizedDescription)
+            }
         }
+        return nil
+    }
+
+    private func sendLoginsStoreErrorToSentry(
+        err: LoginsStoreError,
+        errDomain: String,
+        errorMessage: String
+    ) {
+        var enumVariant: String
+        var errMessage: String
+
+        switch err {
+        case .InvalidRecord(let message):
+            errMessage = message
+            enumVariant = "InvalidRecord"
```

Comment:
If there's a way to dynamically capture the string value of the variant, please let me know. I'd love to not be using raw strings and to reduce the number of lines in this switch statement.

Reply:
Since you are using associated types the message you can get them as a value but not the case itself. 

One way I can suggest is to update the rust component layer as follows: 

<img width="1704" alt="image" src="https://user-images.githubusercontent.com/8919439/183825542-e34b0c43-f2d2-4220-b526-5444881680a4.png">

```
public enum LoginsStorageError {
    // Simple error enums only carry a message
    case InvalidRecord(message: String)

    // Simple error enums only carry a message
    case NoSuchRecord(message: String)

    // Simple error enums only carry a message
    case IncorrectKey(message: String)

    // Simple error enums only carry a message
    case Interrupted(message: String)

    // Simple error enums only carry a message
    case SyncAuthInvalid(message: String)

    // Simple error enums only carry a message
    case RequestFailed(message: String)

    // Simple error enums only carry a message
    case UnexpectedLoginsStorageError(message: String)

    var descriptionValue: String {
        switch self {
        case .InvalidRecord:
            return "InvalidRecord"
        case .NoSuchRecord:
            return "NoSuchRecord"
        case .IncorrectKey:
            return "IncorrectKey"
        case .Interrupted:
            return "Interrupted"
        case .SyncAuthInvalid:
            return "SyncAuthInvalid"
        case .RequestFailed:
            return "RequestFailed"
        case .UnexpectedLoginsStorageError:
            return "UnexpectedLoginsStorageError"
        }
    }
}
```

This will allow you to use it in the function like below: 
```
func printEnumVar(ester: LoginsStorageError) {
    switch ester {
    case .IncorrectKey(let msg):
        print("MSG -> \(msg)")
        print("CASE \(ester.descriptionValue)")
    default:
        print("Default case")
    }
}
```

Reply:
Thanks, @nbhasin2!